### PR TITLE
raftstore: use KvEngine trait instead of implementations

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?
Replace `KvEngine` implementations with `KvEngine` itself in raftstore, so that other applications can use raftstore directly. 

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 
* Use KvEngine trait in raftstore

### CODE REVIEW

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -2729,27 +2740,29 @@ where
...
+pub struct ApplyFsm<EK, ER>
 where
-    E: KvEngine,
+    EK: KvEngine,
+    ER: KvEngine,
 {
-    delegate: ApplyDelegate<E>,
-    receiver: Receiver<Msg<E>>,
-    mailbox: Option<BasicMailbox<ApplyFsm<E>>>,
+    delegate: ApplyDelegate<EK, ER>,
```
Comment (Little-Wallace) [inline components/raftstore/src/store/fsm/apply.rs:2748]: I do not think it is necessary to define a type parameter 'ER' in apply.rs.

File: components/raftstore/src/store/fsm/apply.rs [hunk 2]
```diff
@@ -815,13 +817,16 @@ where
 
     /// The local metrics, and it will be flushed periodically.
     metrics: ApplyMetrics,
+
+    _phantom: PhantomData<ER>,
```
Comment (Little-Wallace) [inline components/raftstore/src/store/fsm/apply.rs:821]: ApplyDelegate does not need 'ER'. 'ER' could be only one type parameter of method which include `ApplyContext` as param.

File: components/raftstore/src/store/fsm/apply.rs [hunk 3]
```diff
@@ -707,22 +703,22 @@ struct WaitSourceMergeState {
...
+    EK: KvEngine,
 {
     /// All of the entries that need to continue to be applied after
     /// the source peer has applied its logs.
     pending_entries: Vec<Entry>,
     /// All of messages that need to continue to be handled after
     /// the source peer has applied its logs and pending entries
     /// are all handled.
-    pending_msgs: Vec<Msg<E>>,
+    pending_msgs: Vec<Msg<EK>>,
```
Comment (Little-Wallace) [inline components/raftstore/src/store/fsm/apply.rs:716]: Is it necessary to replace all 'E' with 'EK'?

**Comment by Little-Wallace** [comment]
rest LGTM

File: components/raftstore/src/store/fsm/apply.rs (hunk 3 above)
Comment (hicqu) [reply components/raftstore/src/store/fsm/apply.rs:716]: I prefer `EK` and `ER` so that in the next PR I can replace all `ER` easilier.

**Pull request merged**